cmd/soroban-cli/src/utils.rs

```diff
@@ -201,62 +204,53 @@ pub fn contract_id_from_str(contract_id: &str) -> Result<[u8; 32], stellar_strke
 /// Might return an error
 pub fn get_contract_spec_from_storage(
     storage: &mut Storage,
-    current_ledger_seq: &u32,
+    _current_ledger_seq: &u32,
     contract_id: [u8; 32],
 ) -> Result<Vec<ScSpecEntry>, FromWasmError> {
     let key = LedgerKey::ContractData(LedgerKeyContractData {
         contract: ScAddress::Contract(contract_id.into()),
         key: ScVal::LedgerKeyContractInstance,
-        body_type: ContractEntryBodyType::DataEntry,
         durability: ContractDataDurability::Persistent,
     });
     match storage.get(&key.into(), &Budget::default()) {
         Ok(rc) => match rc.as_ref() {
             LedgerEntry {
                 data:
                     LedgerEntryData::ContractData(ContractDataEntry {
-                        body:
-                            ContractDataEntryBody::DataEntry(ContractDataEntryData {
-                                val: ScVal::ContractInstance(ScContractInstance { executable, .. }),
-                                ..
-                            }),
-                        expiration_ledger_seq,
+                        val: ScVal::ContractInstance(ScContractInstance { executable, .. }),
                         ..
                     }),
                 ..
             } => match executable {
                 ContractExecutable::Token => {
-                    if expiration_ledger_seq <= current_ledger_seq {
-                        return Err(FromWasmError::NotFound);
-                    }
+                    // TODO: How to identify that an entry is expired now?
+                    //       I(fons) don't think that the storage contains expiration entries and I don't
+                    //       think we can get it from the state in all case (e.g. Token contracts)
+                    // if expiration_ledger_seq <= current_ledger_seq {
+                    //     return Err(FromWasmError::NotFound);
+                    // }
```

Comment:
This is only used for sandbox mode.

Upon reading the code again it seems we should be able to get the spec from the state (which keeps track of expirations) instead of the storage, since the storage is created from the state.

In fact, from what I read, the storage is only initialized and not really filled? I don't know how this could have every worked.

There are no tests so it's hard to just try things without reasoning about it.

Reply:
@willemneal I could use some help here :)

Reply:
@willemneal I've taken a stab at obtaining the spec from the state instead of the storage but I don't think it will work for token contracts. I've also added a couple of (ugly) tests piggybacking on the existing ones, feel free to rewrite them.